components/waypoint-movement.js

```diff
@@ -0,0 +1,239 @@
+
+      this.onFinalWaypointReachedCallbacks = [];
+    },
+    /**
+     * Adds a function to a callback array which, then calls the function after the final waypoint has been reached.
+     * @param  {[type]} f Function to be added.
+     */
+    addOnFinalWaypointReachedCallback: function(f) {
+      this.onFinalWaypointReachedCallbacks.push(f);
+    },
+    /**
+     * Removes a function from the callback array for reaching the final waypoint..
+     * @param  {[type]} f Function to be removed.
+     */
+    removeOnFinalWaypointReachedCallback: function(f) {
+      const index = array.indexOf(f);
+      if (index > -1) {
+        this.onFinalWaypointReachedCallbacks.splice(index, 1);
+      }
+    },
+    /**
+     * Sets the points needed for the next curve
+     */
+     _setCurvePoints: function(){
+      glMatrix.vec3.sub(this.currentPosition, this.positions[this.currentCurveIndex], this.positions[this.currentCurveIndex + 1])
+      glMatrix.vec3.scale(this.currentPosition, this.currentPosition, this.curveDistance);
+      glMatrix.vec3.add(this.p0, this.positions[this.currentCurveIndex + 1], this.currentPosition);
+
+      this.p1 = this.positions[this.currentCurveIndex + 1];
+
+      glMatrix.vec3.sub(this.currentPosition, this.positions[this.currentCurveIndex + 2], this.positions[this.currentCurveIndex + 1])
+      glMatrix.vec3.scale(this.currentPosition, this.currentPosition, this.curveDistance);
+      glMatrix.vec3.add(this.p2, this.positions[this.currentCurveIndex + 1], this.currentPosition);
+    },
+    /**
+     * Sets the object's path object and configures it.
+     * @param  {[WL.Object]} pathObject [Object containing alphebetically named children]
+     */
+    setPathObject: function(pathObject) {
+      this.pathObject = pathObject;
+      this.children = this.pathObject.children.sort(function(a, b) { return a.name.localeCompare(b.name) });
+      this.positions = new Array(this.children.length);
+      for (var i = 0; i < this.children.length; i++) {
+        this.positions[i] = new Float32Array(3);
+        this.children[i].getTranslationWorld(this.positions[i]);
+      }
+      this.currentPositionIndex = 0;
+      this.currentCurveIndex = 0;
+      this.currentLength = 0;
+      this._setCurvePoints();
+
+      this.fromPosition = this.positions[0];
+      this.toPosition = this.positions[1];
+      glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+      this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+    },
+    update: function(dt) {
+      this.currentLength += dt * this.speed;
+      let factor = this.currentLength / this.fromTolength;
+      if(factor > 0.5 && this.currentPositionIndex != this.positions.length - 1) {
+        this.incrementCurveIndex = true;
+      }
+      if(Math.abs(factor - Math.round(factor)) < this.curveDistance && !(this.currentPositionIndex == this.children.length-2 && factor > 0.5) && !(this.currentPositionIndex == 0 && factor < 0.5)) {
```

Comment:
Since this math code is not very transparent, maybe you want to add a short comment of what these if/else branches each do, i.e. "Curve around the point, if +/- curveDistance away from a waypoint" and "Simple straight line if further than curveDistance away" (maybe also a quick explanation on why you are checking children.length)